### PR TITLE
TST: Add test for where inplace

### LINKED ISSUE #22051: where method results in a different outputs when using the inplace kwarg
I am getting different results when using the where method with  `inplace=False` vs `inplace=True`.

consider the following `dataframe`:

```ptyhon 
df = pd.DataFrame({'a':[1,None,3]})
df
    a
0   1.0
1   Nan
2   3.0
```
when using the where method with `inplace=False` (the default), i get the following result:

```python
df.where(pd.notnull(df), None)
    a
0   1.0
1   None
2   3.0
```
with `a` being converted to be of type `object`.

however if i do this:
```python
df.where(pd.notnull(df), None, inplace=True)
df
    a
0   1.0
1   Nan
2   3.0
```
`a` is of type `float64`

pandas version: 0.23.3 , numpy version: 1.14.5

reported because of my [question](https://stackoverflow.com/questions/51515563/pandas-where-method-results-in-a-different-outputs-when-using-the-inplace-kwarg) on SO

### PR DESCRIPTION
**Pull request by rlvoyer**
This adds a test for ensuring that None-to-NaN type-casting is consistent with `where` when `inplace=True` and otherwise.

- [x] closes #22051 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

### CODE REVIEW

File: pandas/tests/frame/indexing/test_where.py [hunk 1]
```diff
@@ -787,3 +791,30 @@ def test_where_columns_casting():
     result = df.where(pd.notnull(df), None)
     # make sure dtypes don't change
     tm.assert_frame_equal(expected, result)
+
+
+optional_ints = st.lists(st.one_of(st.integers(), st.none()), max_size=10, min_size=3)
```
Comment (jreback) [inline pandas/tests/frame/indexing/test_where.py:796]: ideally would want to add lots of other things here, e.g. datetetimes (naive & tz-aware), timedelta, categorical, period, interval. (can be a followup)

also would be better to centralize some of this hypothesis type generation. (do this in this PR)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/common.py [hunk 2]
```diff
@@ -56,3 +58,18 @@ def zip_frames(frames: list[DataFrame], axis: int = 1) -> DataFrame:
         index = frames[0].index
         zipped = [f.loc[i, :] for i in index for f in frames]
         return DataFrame(zipped)
+
+
+OPTIONAL_INTS = st.lists(st.one_of(st.integers(), st.none()), max_size=10, min_size=3)
```
Comment (jreback) [inline pandas/tests/frame/common.py:63]: can you move these to pandas/_testing/_hypothesis.py 

we also have these
```
pandas/tests/io/parser/test_parse_dates.py:from hypothesis import (
pandas/tests/io/parser/test_parse_dates.py:# Strategy for hypothesis
pandas/tests/io/parser/test_parse_dates.py:def _helper_hypothesis_delimited_date(call, date_string, **kwargs):
pandas/tests/io/parser/test_parse_dates.py:def test_hypothesis_delimited_date(date_format, dayfirst, delimiter, test_datetime):
pandas/tests/io/parser/test_parse_dates.py:        except_out_dateutil, result = _helper_hypothesis_delimited_date(
pandas/tests/io/parser/test_parse_dates.py:    except_in_dateutil, expected = _helper_hypothesis_delimited_date(
pandas/tests/tslibs/test_ccalendar.py:from hypothesis import (
pandas/tests/tseries/offsets/test_ticks.py:from hypothesis import (
pandas/tests/tseries/offsets/test_offsets_properties.py:from hypothesis import (
pandas/tests/tseries/offsets/test_offsets_properties.py:from hypothesis.errors import Flaky
pandas/tests/tseries/offsets/test_offsets_properties.py:from hypothesis.extra.dateutil import timezones as dateutil_timezones
pandas/tests/tseries/offsets/test_offsets_properties.py:from hypothesis.extra.pytz import timezones as pytz_timezones
pandas/tests/tseries/offsets/test_offsets_properties.py:        # indicate to hypothesis that this is not a valid test case
```

Comment (jreback) [reply pandas/tests/frame/common.py:63]: e.g. any creation routines i want to move.

**Review by jreback — CHANGES_REQUESTED** [review]
thanks request to move some things around

File: pandas/tests/frame/common.py (hunk 2 above)
Comment (rlvoyer) [reply pandas/tests/frame/common.py:63]: @jreback Just want to make sure I'm clear on what you're asking for here. In addition to the helpers I added, you're asking me to move all hypothesis helpers into that *proposed* new `_hypothesis.py` testing module?

Comment (jreback) [reply pandas/tests/frame/common.py:63]: yes

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rlvoyer

**Pull request merged**